pkg/model/awsmodel/iam.go

```diff
@@ -441,3 +444,41 @@ func (b *IAMModelBuilder) buildAWSIAMRolePolicy(role iam.Subject) (fi.Resource,
 
 	return fi.NewStringResource(policy), nil
 }
+
+func (b *IAMModelBuilder) FindDeletions(context *fi.ModelBuilderContext, cloud fi.Cloud) error {
```

Comment:
Is model the right place to implement this?

Reply:
I think this is no less awkward than what I did with SG rules.
The challenge is that there is no relationship between tasks and models. Another model could also provision iam roles if that made sense.

An easy solution to this would be that the model add an additional tag to the resources it creates. And then replace ownershipTag below with that tag. E.g `kops.k8s.io/<cluster name>/owner: awsmodel/iammodelbuilder`

Reply:
> I think this is no less awkward than what I did with SG rules.

I think https://github.com/kubernetes/kops/pull/10251 is not yet merged.

> The challenge is that there is no relationship between tasks and models.

It's always a challenge to track and handle orphaned resources and various hacks were made for that reason. I think the main reason this was not made easier initially was the lack of tags at that time. Another reason is that we 
Moving this logic to the model not such a good idea  and if we don't find a good alternative now will become a permanent thing.

Reply:
This is absolutely the right place to implement this. The model knows what is supposed to be there, so after it has constructed what should be there it can compare that against what actually is there. Just like any well-behaved controller.

One could talk about consolidating the cluster and instancegroup destruction code into this mechanism in order to reduce the possibility of logic discrepancies, but that would be for future PRs.

Reply:
> The challenge is that there is no relationship between tasks and models.

There is a relationship: models create tasks. This PR uses the tasks generated by `Loader.BuildTasks()` as a record of which objects the model says should exist. This is why `FindDeletions` runs as a second pass.

Reply:
I was not opposed to this, just not a fan of adding AWS SDK code in model.
Also, the FindDeletions name is something that I would see more as FindOrphaned and leave FindDeletions for tasks.
In general, making use of this may fix some of the awkwardness in some task deletions.

Reply:
Well, this particular `ModelBuilder` is AWS-specific and in a directory for AWS-specific model code.

I made `HasDeletions` require `ModelBuilder` so that both could be added to `Loader.Builders`. I wanted to avoid errors where something that was both didn't get added to both lists.

But `HasDeletions` is closer to `Task` than `ModelBuilder`, despite needing to take a `ModelBuilderContext` in order to construct new `Task`s.